### DISCUSSION TITLE
Hoisting vs peerDependencies

### DISCUSSION THREAD

**Discussion by jeffrson** [yarnpkg/berry#2754] [Q&A] [answered]
Hi,
in my workspace, there is a library that depends on "fomantic-ui".

"fomantic-ui" cannot be installed with yarn/yarn2/npm7, but that's another story. Once, I installed it with npm6 and it works inside its own module.

With yarn1, we had a "nohoist" entry in the root of the workspace:
```
  "nohoist": [
      "**/fomantic-ui"
    ]
```
which worked quite flawlessly. However, "nohoist" is deprecated now (I read how it was related to React Native a.s.o.).

The new
```
    "installConfig": {
        "hoistingLimits": "workspaces"
    }
```

on the one hand prevents hoisting of *all* dependencies of the respective package. Not too good, but bearable.

On the other hand, if there's a "peerDependency" inside this same new package.json all depending packages cause "fomantic-ui" hoisted to the root of the workspace as well as being kept locally. Unfortunately the package does not work inside the workspace root (which is the very reason for limiting it being hoisted).
If I put the "installConfig" entry into all those depending packages they all will keep their dependencies un-hoisted. Which will increase required space for those packages and, most notably, will largely increase the time for running "yarn".

Did I something wrong? Or might it be a bug in yarn (2.4.1)? Are there any recommendations?

BTW, with pnp, hoisting is always to the root of the workspace, isn't it? That again will be a problem with fomantic-ui (since it will not work there...).

**Comment by larixer** [comment, marked as answer]
To prevent hoisting only a few packages, there is a workaround - you can keep the name busy at the upper-level, for example by using `link:` protocol, or by using an alias. In the root `package.json` you can have:
```json
{
  "name": "foo",
  "workspaces": [
    "packages/*"
  ],
  "dependencies": {
    "fomantic-ui": "link:./packages"
  }
}
```
This will prevent hoisting `fomantic-ui` from workspaces to the top level. It is just an example, I'm showing you only the idea of a workaround.

**Reply by jeffrson** [reply]
Yes, indeed. This does the trick! Thank you!

**Comment by merceyz** [comment]
> Did I something wrong? Or might it be a bug in yarn (2.4.1)? Are there any recommendations?

No, it seems like everything is behaving as intended. There was work done to improve `portal:` support in the node-modules linker in https://github.com/yarnpkg/berry/pull/2481, there are plans to expand it to work with workspaces as well so you don't have to add `hoistingLimits: workspaces` to all workspaces.

> BTW, with pnp, hoisting is always to the root of the workspace, isn't it? That again will be a problem with fomantic-ui (since it will not work there...).

With PnP there is no hoisting as the layout on disk doesn't matter, it just points to packages in the cache which gives you [perfect deduplication](https://yarnpkg.com/features/pnp#fixing-node_modules)

**Reply by jeffrson** [reply]
fomantic-ui is special, unfortunately - it tries to find a file (semantic.json) next to the nearest package.json, which apparently is in workspaces root when using pnp. With node-modules linker, without hoisting, fomantic-ui loads semantic.json correctly from the workspace itself.

Don't know if this makes sense, but this is how it seems to behave at the moment.
